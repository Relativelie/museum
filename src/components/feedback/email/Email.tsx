import { useState, FC, SyntheticEvent, useEffect, ChangeEvent } from 'react';

export const Email: FC = () => {
    const [rule] = useState(/^[^\s@]+@[^\s@]+\.[^\s@]+$/);
    const [isCorrect, setIsCorrect] = useState(true);
    const [incorrectClass, setIncorrectClass] = useState('');
    const [inputValue, setInputValue] = useState('');

    useEffect(() => {
        if (!isCorrect) setIncorrectClass('incorrectInputValue');
        else setIncorrectClass('');
    });

    const checkEmail = (eventType: string, event: SyntheticEvent<EventTarget>) => {
        if (eventType === 'blur' || eventType === 'Enter') {
            const elem = event.target as HTMLInputElement;
            setIsCorrect(rule.test(elem.value));
            elem.blur();
        } else return;
    };

    const changeEmailValue = (event: ChangeEvent<HTMLInputElement>) => {
        setInputValue(event.target.value.trim());
    };

    return (
        <input
            className={`feedback_input ${incorrectClass}`}
            name="email"
            type="email"
            data-testid="form-field-email"
            placeholder="Email"
            value={inputValue}
            onChange={(e) => changeEmailValue(e)}
            onBlur={(e) => checkEmail(e.type, e)}
            onKeyUp={(e) => checkEmail(e.key, e)}
        />
    );
};
