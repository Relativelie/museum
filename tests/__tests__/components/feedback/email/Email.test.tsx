import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import { act, create } from 'react-test-renderer';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { fireEvent, screen } from '@testing-library/dom';

import { Email } from '../../../../../src/components/feedback/email/Email';

describe('Feedback - Email component', () => {
    test('snapshot - email component', () => {
        let root: any;
        act(() => {
            root = create(<Email />);
        });
        expect(root.toJSON()).toMatchSnapshot();
    });

    test('change email value with enter press - correct value', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'test@test.test');
        expect(inputEl).toHaveValue('test@test.test');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 13 });
        expect(inputEl).toHaveClass('feedback_input');
    });

    test('change email value with blur activate - correct value', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'correct.@correct.ty');
        expect(inputEl).toHaveValue('correct.@correct.ty');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input');
    });

    test('change email value with enter press - incorrect value: without part of domain name', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'test@test.');
        expect(inputEl).toHaveValue('test@test.');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 13 });
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('change email value with blur activate - incorrect value: without domain name', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'test@');
        expect(inputEl).toHaveValue('test@');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('change email value with enter press - incorrect value: without unique user name', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, '@test.test');
        expect(inputEl).toHaveValue('@test.test');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 'Enter', charCode: 0 });
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('change email value with blur activate - incorrect value: without at sign', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'testtest.test');
        expect(inputEl).toHaveValue('testtest.test');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('change email value with blur activate - empty value', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.click(inputEl);
        expect(inputEl).toHaveValue('');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('change email value with enter press - incorrect value: only spaces', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, '     ');
        expect(inputEl).toHaveValue('');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 'Enter', charCode: 0 });
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });

    test('add incorrect value, then add correct value', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'freef');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 13 });
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
        userEvent.clear(inputEl);
        userEvent.type(inputEl, 'test@test.test');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input');
    });

    test('add correct value, then add incorrect value', () => {
        render(<Email />);
        const inputEl = screen.getByTestId('form-field-email');
        userEvent.type(inputEl, 'test@test.test');
        userEvent.tab();
        expect(inputEl).toHaveClass('feedback_input');
        userEvent.clear(inputEl);
        userEvent.type(inputEl, 'f');
        fireEvent.keyUp(inputEl, { key: 'Enter', code: 13 });
        expect(inputEl).toHaveClass('feedback_input incorrectInputValue');
    });
});
