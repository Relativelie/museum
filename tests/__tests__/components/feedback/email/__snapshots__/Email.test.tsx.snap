// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback - Email component snapshot - email component 1`] = `
<input
  className="feedback_input "
  data-testid="form-field-email"
  name="email"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyUp={[Function]}
  placeholder="Email"
  type="email"
  value=""
/>
`;
